Sources/NIOCore/AsyncSequences/NIOAsyncSequenceProducer.swift

```diff
@@ -0,0 +1,1064 @@
+///
+/// - Important: This sequence is a unicast sequence that only supports a single ``NIOAsyncSequenceProducer/AsyncIterator``.
+/// If you try to create more than one iterator it will result in a `fatalError`.
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+public struct NIOAsyncSequenceProducer<
+    Element: Sendable,
+    Strategy: NIOAsyncSequenceProducerBackPressureStrategy,
+    Delegate: NIOAsyncSequenceProducerDelegate
+>: Sendable {
+    /// Simple struct for the return type of ``NIOAsyncSequenceProducer/makeSourceAndSequence(backPressureStrategy:delegate:)``.
+    public struct NewSequence {
+        /// The source of the ``NIOAsyncSequenceProducer`` used to yield and finish.
+        public let source: Source
+        /// The actual sequence which should be passed to the consumer.
+        public let sequence: NIOAsyncSequenceProducer
+
+        @usableFromInline
+        /* fileprivate */ internal init(
+            source: Source,
+            sequence: NIOAsyncSequenceProducer
+        ) {
+            self.source = source
+            self.sequence = sequence
+        }
+    }
+
+    /// This class is needed to hook the deinit to observe once all references to the ``NIOAsyncSequenceProducer`` are dropped.
+    ///
+    /// If we get move-only types we should be able to drop this class and use the `deinit` of the ``AsyncIterator`` struct itself.
+    ///
+    /// - Important: This is safe to be unchecked ``Sendable`` since the `storage` is ``Sendable`` and `immutable`.
+    @usableFromInline
+    /* fileprivate */ internal final class InternalClass: @unchecked NIOSendable {
+        @usableFromInline
+        internal let storage: Storage
+
+        @inlinable
+        init(storage: Storage) {
+            self.storage = storage
+        }
+
+        @inlinable
+        deinit {
+            storage.sequenceDeinitialized()
+        }
+    }
+
+    @usableFromInline
+    /* private */ internal let internalClass: InternalClass
+
+    @usableFromInline
+    /* private */ internal var storage: Storage {
+        self.internalClass.storage
+    }
+
+    /// Initializes a new ``NIOAsyncSequenceProducer`` and a ``NIOAsyncSequenceProducer/Source``.
+    ///
+    /// - Important: This method returns a tuple containing a ``NIOAsyncSequenceProducer/Source`` and
+    /// a ``NIOAsyncSequenceProducer``. The source MUST be held by the caller and
+    /// used to signal new elements or finish. The sequence MUST be passed to the actual subscriber and MUST NOT be held by the
+    /// caller. This is due to the fact that deiniting the sequence is used as part of a trigger to terminate the underlying source.
+    ///
+    /// - Parameters:
+    ///   - backPressureStrategy: The back-pressure strategy of the sequence.
+    ///   - delegate: The delegate of the sequence
+    /// - Returns: A ``NIOAsyncSequenceProducer/Source`` and a ``NIOAsyncSequenceProducer``.
+    @inlinable
+    public static func makeSourceAndSequence(
```

Comment:
The call sites are a bit rough but could be made simpler rather easily:

```
        let result = NIOAsyncSequenceProducer<
             Int,
             MockNIOElementStreamBackPressureStrategy,
             MockNIOBackPressuredStreamSourceDelegate
         >.makeSourceAndSequence(
             backPressureStrategy: self.backPressureStrategy,
             delegate: self.delegate
         )
```

via:

```
    @inlinable
     public static func makeSourceAndSequence(
        of elementType: Element.Type = Element.self, 
        backPressureStrategy: Strategy,
         delegate: Delegate
     ) -> NewSequence {
```

so it is spelled more nicer on callsites:

```
NIOAsyncSequenceProducer.makeSequence(
             of: Int.self
             backPressureStrategy: self.backPressureStrategy,
             delegate: self.delegate)
```

not sure the "AND" in `SourceAndSequence` explains all that much, I'd consider dropping... In general always worried when I see APIs mention "X AND Y".

Reply:
Great suggestion! Changed it around, feels way better to create a sequence now.